securedrop/journalist_app/api2/events.py

```diff
@@ -0,0 +1,151 @@
+from db import db
+from journalist_app import utils
+from journalist_app.api2.shared import save_reply
+from journalist_app.api2.types import (
+    Event,
+    EventResult,
+    EventStatusCode,
+    EventType,
+    ItemTarget,
+    SourceTarget,
+)
+from journalist_app.sessions import Session
+from models import Reply, Source, Submission
+from redis import Redis
+from sqlalchemy.orm.exc import MultipleResultsFound, NoResultFound
+
+# `IDEMPOTENCE_PERIOD` MUST be greater than or equal to
+# `sdconfig.SecureDropConfig.SESSION_LIFETIME`.  In practice, 24 hours is the
+# easiest period to reason about.
+IDEMPOTENCE_PERIOD = 60 * 60 * 24  # seconds * minutes * hours = 1 day
+
+REDIS_EVENT_PREFIX = "sd/events"
+
+
+class EventHandler:
+    """
+    This class is the per-request context for handling events.  To add a handler
+    for a new event `thing_done`, you must:
+
+    1. define the enum value `EventType.THING_DONE` in journalist_api2.types;
+
+    2. define the handler as a static method `handle_thing_done(event: Event)`
+       in this class
+
+    3. explicitly register `{"thing_done": self.handle_thing_done}` inside
+      `EventHandler.process()`.
+
+    This is belt-and-suspenders for ensuring that only the intended methods are
+    exposed as callable event handlers.
+    """
+
+    def __init__(self, session: Session, redis: Redis) -> None:
+        self._session = session
+        self._redis = redis
+
+    def process(self, event: Event) -> EventResult:
+        """The per-event entry-point for handling a single event."""
+
+        try:
+            event.type = EventType(event.type)  # strict enum
```

Comment:
I think we should move this validation logic to also be pre-processing, that way regardless of where we are, once we start dealing with `Event` objects we know they're already validated and using the correct internal types. Or to put it another way, we should only have one place that can raise TypeErrors.

So maybe we have a `deserialize_events_from_user_input` function that takes care of this, or integrate this into Event's [https://docs.python.org/3/library/dataclasses.html#dataclasses.__post_init__](`__post_init__`)?